Ubuntu/History/historymanager.cpp

```diff
@@ -0,0 +1,98 @@
+ *
+ * This file is part of history-service.
+ *
+ * history-service is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * history-service is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "historymanager.h"
+#include "eventview.h"
+#include "manager.h"
+#include <QDebug>
+#include <QJSValueList>
+#include <QTimer>
+
+
+HistoryManager::HistoryManager(QObject *parent) :
+    QObject(parent), mPendingOperation(false)
+{
+}
+
+void HistoryManager::removeEvents(int eventType, const QString &from, const QJSValue &callback)
+{
+    if (!callback.isCallable()) {
+        qCritical() << "no callback found!";
+        return;
+    }
+
+    QJSValue result(callback);
+    if (mPendingOperation) {
+        result.call(QJSValueList { OperationError::OPERATION_ALREADY_PENDING });
+        qWarning() << "there is a pending operation, request cancelled";
+        return;
+    }
+
+    QDateTime fromDate = QDateTime::fromString(from, Qt::ISODate);
+    History::EventType type = (History::EventType) eventType;
+
+    if (type == History::EventTypeNull|| !fromDate.isValid()) {
+        result.call(QJSValueList { OperationError::OPERATION_INVALID });
+        qWarning() << "invalid type or date, request cancelled";
+        return;
+    }
+
+    History::Filter queryFilter(History::FieldTimestamp, QVariant(from), History::MatchLess);
+    History::Sort querySort(History::FieldTimestamp, Qt::DescendingOrder);
+
+    if (!queryFilter.isValid()) {
+        result.call(QJSValueList { OperationError::OPERATION_INVALID });
+        qWarning() << "invalid filter, operation cancelled";
+        return;
+    }
+
+    if (!mView.isNull()) {
+        mView->disconnect(this);
+    }
+
+    int eventsCount = History::Manager::instance()->eventsCount(type, queryFilter);
+
+    if (eventsCount > 0) {
+        mPendingOperation = true;
+        int deletedCount = 0;
+        mView = History::Manager::instance()->queryEvents(type, querySort, queryFilter);
+
+        connect(mView.data(), &History::EventView::eventsRemoved,
+                [this, &deletedCount, eventsCount, callback]( const History::Events &events ) {
+
+            deletedCount += events.count();
+            if (deletedCount == eventsCount) {
+                mPendingOperation = false;
+            }
+            QJSValue result(callback);
+            result.call(QJSValueList { OperationError::NO_ERROR, eventsCount, deletedCount });
+        });
+
+        History::Manager::instance()->removeEvents(type, queryFilter, querySort);
```

Comment:
Do we have an "error" or "done" signal in the view, that we could connect to, instead of having a timeout, in order to know when the operation is completed?

Reply:
You are right, i don't think it is needed. The " ok all finished" can only be done from our side because component use pagination and trigger s "removeEvents" signal each time a batch is finished.
But just found "invalidated" signal that can occur when something went wrong, i need to connect to it in case